polar-core/src/data_filtering.rs

```diff
@@ -63,737 +67,632 @@ pub struct FetchRequest {
 // @Q(steve): Is it always the last one in the resolve_order?
 #[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
 pub struct ResultSet {
-    requests: HashMap<String, FetchRequest>,
-    resolve_order: Vec<String>,
-    result_id: String,
+    requests: HashMap<Id, FetchRequest>,
+    resolve_order: Vec<Id>,
+    result_id: Id,
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Default)]
 pub struct FilterPlan {
     result_sets: Vec<ResultSet>,
 }
 
-impl FilterPlan {
-    pub fn explain(&self) {
-        eprintln!("UNION");
-        for (i, result_set) in self.result_sets.iter().enumerate() {
-            eprintln!("  =Result Set: {}=", i);
-            for id in &result_set.resolve_order {
-                let fetch_request = result_set.requests.get(id).unwrap();
-                eprintln!("    {}: Fetch {}", id, fetch_request.class_tag);
-                for constraint in &fetch_request.constraints {
-                    let op = match constraint.kind {
-                        ConstraintKind::Eq => "=".to_owned(),
-                        ConstraintKind::In => "in".to_owned(),
-                        ConstraintKind::Contains => "contains".to_owned(),
-                    };
-                    let field = constraint.field.clone();
-                    let value = match &constraint.value {
-                        ConstraintValue::Term(t) => t.to_polar(),
-                        ConstraintValue::Field(f) => format!("FIELD({})", f),
-                        ConstraintValue::Ref(r) => {
-                            let mut s = "REF(".to_owned();
-                            if let Some(field) = &r.field {
-                                s.push_str(&format!("field {} of ", field));
-                            }
-                            s.push_str(&format!("result {})", r.result_id));
-                            s
-                        }
-                    };
-                    eprintln!("          {} {} {}", field, op, value);
-                }
-            }
-        }
-    }
-}
-
 pub type Types = HashMap<String, HashMap<String, Type>>;
 pub type PartialResults = Vec<ResultEvent>;
 
-#[derive(Debug)]
+#[derive(Debug, Default)]
 struct VarInfo {
     cycles: Vec<(Symbol, Symbol)>,                      // x = y
     types: Vec<(Symbol, String)>,                       // x matches XClass
     eq_values: Vec<(Symbol, Term)>,                     // x = 1;
     contained_values: Vec<(Term, Symbol)>,              // 1 in x
     field_relationships: Vec<(Symbol, String, Symbol)>, // x.a = y
     in_relationships: Vec<(Symbol, Symbol)>,            // x in y
+    counter: Counter,
+}
+
+#[derive(Debug)]
+struct Vars {
+    variables: HashMap<Id, HashSet<Symbol>>,
```

Comment:
Dave requested comments, similar to the ones on VarInfo, that explain what each of these are.